File: tyhp-lang/vscode/src/lsp/restartBackoff.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import {
    DEFAULT_RESTART_INITIAL_MS,
    DEFAULT_RESTART_MAX_MS,
    RestartBackoff,
    nextRestartDelayMs,
    shouldScheduleCrashRestart,
} from "./restartBackoff";

test("first crash uses the initial delay", () => {
    assert.equal(nextRestartDelayMs(1), DEFAULT_RESTART_INITIAL_MS);
    assert.equal(nextRestartDelayMs(0), DEFAULT_RESTART_INITIAL_MS);
});

test("delays double until the cap", () => {
    assert.equal(nextRestartDelayMs(2), 2_000);
    assert.equal(nextRestartDelayMs(3), 4_000);
    assert.equal(nextRestartDelayMs(4), 8_000);
    assert.equal(nextRestartDelayMs(5), 16_000);
    assert.equal(nextRestartDelayMs(6), DEFAULT_RESTART_MAX_MS);
    assert.equal(nextRestartDelayMs(20), DEFAULT_RESTART_MAX_MS);
});

test("RestartBackoff increments then resets after a healthy start", () => {
    const backoff = new RestartBackoff(100, 800);
    assert.equal(backoff.nextDelayMs(), 100);
    assert.equal(backoff.nextDelayMs(), 200);
    assert.equal(backoff.nextDelayMs(), 400);
    assert.equal(backoff.nextDelayMs(), 800);
    assert.equal(backoff.nextDelayMs(), 800);
    assert.equal(backoff.consecutiveFailures, 5);
    backoff.reset();
    assert.equal(backoff.consecutiveFailures, 0);
    assert.equal(backoff.nextDelayMs(), 100);
});

test("a server that never reached Running is not restarted", () => {
    assert.equal(
        shouldScheduleCrashRestart({ neverStarted: true, consecutiveFailures: 0 }),
        false
    );
    assert.equal(
        shouldScheduleCrashRestart({ neverStarted: true, consecutiveFailures: 2 }),
        false
    );
});

test("a running server is retried until the consecutive-failure cap", () => {
    assert.equal(
        shouldScheduleCrashRestart({ neverStarted: false, consecutiveFailures: 0 }),
        true
    );
    assert.equal(
        shouldScheduleCrashRestart({ neverStarted: false, consecutiveFailures: 2 }),
        true
    );
    assert.equal(
        shouldScheduleCrashRestart({ neverStarted: false, consecutiveFailures: 3 }),
        false
    );
    assert.equal(
        shouldScheduleCrashRestart({
            neverStarted: false,
            consecutiveFailures: 1,
            maxFailures: 1,
        }),
        false
    );
});

File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/lsp/RestartBackoffTest.kt
package com.tyhp.lang.lsp

import kotlin.test.Test
import kotlin.test.assertEquals

class RestartBackoffTest {
    @Test
    fun `first crash uses the initial delay`() {
        assertEquals(DEFAULT_RESTART_INITIAL_MS, nextRestartDelayMs(1))
        assertEquals(DEFAULT_RESTART_INITIAL_MS, nextRestartDelayMs(0))
    }

    @Test
    fun `delays double until the cap`() {
        assertEquals(2_000L, nextRestartDelayMs(2))
        assertEquals(4_000L, nextRestartDelayMs(3))
        assertEquals(8_000L, nextRestartDelayMs(4))
        assertEquals(16_000L, nextRestartDelayMs(5))
        assertEquals(DEFAULT_RESTART_MAX_MS, nextRestartDelayMs(6))
        assertEquals(DEFAULT_RESTART_MAX_MS, nextRestartDelayMs(20))
    }

    @Test
    fun `RestartBackoff increments then resets after a healthy start`() {
        val backoff = RestartBackoff(100, 800)
        assertEquals(100L, backoff.nextDelayMs())
        assertEquals(200L, backoff.nextDelayMs())
        assertEquals(400L, backoff.nextDelayMs())
        assertEquals(800L, backoff.nextDelayMs())
        assertEquals(800L, backoff.nextDelayMs())
        assertEquals(5, backoff.consecutiveFailures)
        backoff.reset()
        assertEquals(0, backoff.consecutiveFailures)
        assertEquals(100L, backoff.nextDelayMs())
    }
}
